cuda/matrix/hybrid_kernels.cu

```diff
@@ -0,0 +1,436 @@
+            }
+            temp_val = 0;
+        }
+        curr_row = next_row;
+    }
+    if (num > 0) {
+        ind = ind_start + (num - 1) * subwarp_size;
+        temp_val += ind < nnz && val[ind] != zero<ValueType>();
+        // segmented scan
+        is_first_in_segment = true;
+        segment_scan<subwarp_size>(tile_block, curr_row, &temp_val,
+                                   &is_first_in_segment);
+        if (is_first_in_segment) {
+            atomic_add(&(c[curr_row]), temp_val);
+        }
+    }
+}
+
+
+template <typename ValueType>
+__global__ __launch_bounds__(default_block_size) void reduce_nnz(
+    size_type size, const ValueType *__restrict__ nnz_per_row,
+    ValueType *__restrict__ result)
+{
+    extern __shared__ ValueType block_sum[];
+    reduce_array(size, nnz_per_row, block_sum,
+                 [](const ValueType &x, const ValueType &y) { return x + y; });
+
+    if (threadIdx.x == 0) {
+        result[blockIdx.x] = block_sum[0];
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(default_block_size) void fill_in_csr(
+    size_type num_rows, size_type max_nnz_per_row, size_type stride,
+    const ValueType *__restrict__ ell_val,
+    const IndexType *__restrict__ ell_col,
+    const ValueType *__restrict__ coo_val,
+    const IndexType *__restrict__ coo_col,
+    const IndexType *__restrict__ coo_offset,
+    IndexType *__restrict__ result_row_ptrs,
+    IndexType *__restrict__ result_col_idxs,
+    ValueType *__restrict__ result_values)
+{
+    const auto tidx = threadIdx.x + blockDim.x * blockIdx.x;
+
+    if (tidx < num_rows) {
+        auto write_to = result_row_ptrs[tidx];
+        for (auto i = 0; i < max_nnz_per_row; i++) {
+            const auto source_idx = tidx + stride * i;
+            if (ell_val[source_idx] != zero<ValueType>()) {
+                result_values[write_to] = ell_val[source_idx];
+                result_col_idxs[write_to] = ell_col[source_idx];
+                write_to++;
+            }
+        }
+        for (auto i = coo_offset[tidx]; i < coo_offset[tidx + 1]; i++) {
+            if (coo_val[i] != zero<ValueType>()) {
+                result_values[write_to] = coo_val[i];
+                result_col_idxs[write_to] = coo_col[i];
+                write_to++;
+            }
+        }
+    }
+}
+
+
+template <typename ValueType1, typename ValueType2>
+__global__ __launch_bounds__(default_block_size) void add(
+    size_type num, ValueType1 *__restrict__ val1,
+    const ValueType2 *__restrict__ val2)
+{
+    const auto tidx = threadIdx.x + blockDim.x * blockIdx.x;
+    if (tidx < num) {
+        val1[tidx] += val2[tidx];
+    }
+}
+
+
+// Copied from ell_kernel.cu
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(default_block_size) void count_ell_nnz_per_row(
+    size_type num_rows, size_type max_nnz_per_row, size_type stride,
+    const ValueType *__restrict__ values, IndexType *__restrict__ result)
```

Comment:
You should not have to copy this. You can make a kernel interface for this in ell, include the correct header and then directly call `ell:count_nnz_per_row(...);`